### PR TITLE
feat: AdjustedBorrow and InterestScalar

### PR DESCRIPTION
**Pull request by toteki**
## Description

Main changes are as follows:
- Proto files changed, due to `GenesisState` changes and the dropping of `InterestEpoch`
- Changes to how certain quantities are derived (Borrow Amount, Total Borrows, Borrow Utilization, Borrow APY, Lend APY, Exchange Rates, Interest Accrued) and resulting effects on functions that use them
- Tests modified slightly as we can no longer set the above values arbitrarily
- Iterators simplifies to use the `k.iterate` helper made by @RafilxTenfen
- Spec update

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [x] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/leverage/spec/01_concepts.md [hunk 1]
```diff
@@ -65,13 +65,21 @@ As a reminder, the following values are always available as a basis for calculat
 - Total supply of any token or uToken denomination, available through the `bank` module.
 - The `leverage` module account balance, available through the `bank` module.
 - Collateral uToken amounts held in the `leverage` module account for individual borrowers, stored in `leverage` module [State](02_state.md).
-- Borrowed denominations and amounts for individual borrowers, stored in `leverage` module [State](02_state.md).
-- Total borrows summed over all borrower accounts, derived from the above.
+- Borrowed denominations and _adjusted amounts_ for individual borrowers, stored in `leverage` module [State](02_state.md).
```
Comment (toteki) [inline x/leverage/spec/01_concepts.md:68]:
```suggestion
- Borrowed denominations and _adjusted amounts_ for individual borrowers, stored in `leverage` module [State](02_state.md).
- _Interest scalars_ for all borrowed denominations, which are used with adjusted borrow amounts
```

File: x/leverage/spec/05_endblock.md [hunk 2]
```diff
@@ -1,8 +1,6 @@
 # End Block
 
-At the end of every block, the leverage module checks whether the current block height is a multiple of `InterestEpoch`.
-
-Every `InterestEpoch`, it runs the following steps in order:
+Every block, the leverage module runs the following steps in order:
 - Repay bad debts using reserves
 - Update borrow and lend APY and accrue interest on borrows
 - Update uToken exchange rates
```
Comment (toteki) [inline x/leverage/spec/05_endblock.md:6]:
```suggestion
- Accrue interest on borrows
```

File: x/leverage/spec/06_events.md [hunk 3]
```diff
@@ -70,28 +70,28 @@ The leverage module emits the following events:
...
-| interest_epoch | reserved       | {newReserves}          |
+| Type             | Attribute Key  | Attribute Value        |
+| ---------------- | -------------- | ---------------------- |
+| interest_accrual | block_height   | {ctx.BlockHeight}      |
+| interest_accrual | unix_time      | {ctx.BlockTime.Unix()} |
+| interest_accrual | total_interest | {totalInterest}        |
+| interest_accrual | reserved       | {newReserves}          |
 
-Interest epochs emit an event with their current block height and time, as well as total interest accrued across all borrows and the amount of each token added to reserves. Occurs every `InterestEpoch`.
+Interest epochs emit an event with their current block height and time, as well as total interest accrued across all borrows and the amount of each token added to reserves. Occurs every block.
```
Comment (toteki) [inline x/leverage/spec/06_events.md:86]:
```suggestion
Interest accrual emits an event with the current block height and time, as well as total interest accrued across all borrows and the amount of each token added to reserves. Occurs every block.
```

File: x/leverage/keeper/keeper_test.go [hunk 4]
```diff
@@ -175,48 +175,6 @@ func (s *IntegrationTestSuite) TestWithdrawAsset_Valid() {
 	s.Require().Equal(int64(0), uTokenBalance.Amount.Int64())
 }
 
-func (s *IntegrationTestSuite) TestWithdrawAsset_WithExchangeRate() {
```
Comment (toteki) [inline x/leverage/keeper/keeper_test.go:178]: Note: Removed test as manually setting exchange rates is no longer possible (they are always derived instantly)

File: x/leverage/keeper/keeper_test.go [hunk 5]
```diff
@@ -231,22 +189,6 @@ func (s *IntegrationTestSuite) TestSetReserves() {
 	s.Require().Equal(amount, sdk.NewInt(200000000))
 }
 
-func (s *IntegrationTestSuite) TestSetExchangeRate() {
```
Comment (toteki) [inline x/leverage/keeper/keeper_test.go:234]: Note: Removed test as manually setting exchange rates is no longer possible (they are always derived instantly)

File: x/leverage/keeper/keeper_test.go [hunk 6]
```diff
@@ -1193,29 +1131,6 @@ func (s *IntegrationTestSuite) TestGetEligibleLiquidationTargets_TwoAddr() {
 	s.Require().Equal([]sdk.AccAddress{lenderAddr, anotherLender}, lenderAddress)
 }
 
-func (s *IntegrationTestSuite) TestExchangeRatesInvariant() {
```
Comment (toteki) [inline x/leverage/keeper/keeper_test.go:1196]: Note: Removed test as manually setting exchange rates is no longer possible (they are always derived instantly)

File: x/leverage/keeper/keeper_test.go [hunk 7]
```diff
@@ -1273,36 +1188,6 @@ func (s *IntegrationTestSuite) TestBorrowAmountInvariant() {
 	s.Require().False(broken)
 }
 
-func (s *IntegrationTestSuite) TestBorrowAPYInvariant() {
```
Comment (toteki) [inline x/leverage/keeper/keeper_test.go:1276]: Note: Removed test as manually setting borrow APY is no longer possible (always derived instantly)

File: x/leverage/keeper/keeper_test.go [hunk 8]
```diff
@@ -1273,36 +1188,6 @@ func (s *IntegrationTestSuite) TestBorrowAmountInvariant() {
...
-
-	// sets the borrow APY to 0
-	s.app.LeverageKeeper.SetBorrowAPY(s.ctx, umeeapp.BondDenom, sdk.NewDec(0))
-
-	// check invariant
-	_, broken = keeper.BorrowAPYInvariant(s.app.LeverageKeeper)(s.ctx)
-	s.Require().False(broken)
-}
-
-func (s *IntegrationTestSuite) TestLendAPYInvariant() {
```
Comment (toteki) [inline x/leverage/keeper/keeper_test.go:1291]: Note: Removed test as manually setting lend APY is no longer possible (always derived instantly)

File: x/leverage/keeper/store.go [hunk 9]
```diff
@@ -0,0 +1,130 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/umee-network/umee/x/leverage/types"
+)
+
```
Comment (toteki) [inline x/leverage/keeper/store.go:9]: This `store.go` file only contains the new un-exported get/set functions that were made for this feature, but we might consider moving more of them here in a separate PR.

File: x/leverage/keeper/iter.go [hunk 10]
```diff
@@ -0,0 +1,228 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/umee-network/umee/x/leverage/types"
+)
+
```
Comment (toteki) [inline x/leverage/keeper/iter.go:8]: I have refactored all iterators in `x/leverage` to use the `k.iterate` helper which @RafilxTenfen made.

They have been moved to this file, with the exception of the ones that export genesis-related structs (those are in genesis.go).

File: x/leverage/keeper/token.go [hunk 11]
```diff
@@ -58,22 +58,14 @@ func (k Keeper) SetRegisteredToken(ctx sdk.Context, token types.Token) {
 		panic(fmt.Sprintf("failed to encode token: %s", err))
 	}
 
-	// for tokens not previously registered, set tokens:uToken to 1.0
-	if err := k.InitializeExchangeRate(ctx, token.BaseDenom); err != nil {
```
Comment (toteki) [inline x/leverage/keeper/token.go:62]: Utoken exchange rates are derived instantly, so initializing is no longer required

**Review by toteki — COMMENTED** [review]
PR is still fairly cluttered after being done a second time, but it's mostly a result of the changes to how a bunch of quantities are derived and stored now.

File: x/leverage/keeper/borrows.go [hunk 12]
```diff
@@ -12,152 +11,70 @@ import (
...
 	if bz := store.Get(key); bz != nil {
-		err := owed.Amount.Unmarshal(bz)
+		err := adjustedAmount.Unmarshal(bz)
 		if err != nil {
 			panic(err)
 		}
 	}
 
+	// Apply interest scalar
+	owed.Amount = adjustedAmount.Mul(k.getInterestScalar(ctx, denom)).TruncateInt()
```
Comment (alexanderbez) [inline x/leverage/keeper/borrows.go:25]: Is truncating a concern here? Meaning, would rounding inefficiencies yield any potential issues?

**Review by alexanderbez — APPROVED** [review]
Excellent work @toteki. I couldn't catch any technical issues with the PR and spec updates look reasonable as well.

My only question/doubt is the truncation we do for deriving adjusted (total) borrows. Would that lead to any issues?

File: x/leverage/keeper/borrows.go (hunk 12 above)
Comment (toteki) [reply x/leverage/keeper/borrows.go:25]: I'll look into it - especially in cases where a repayment or some other operation might leave tiny borrowed amount remainders instead of zero, since that could cause issues withdrawing collateral for example.

I suspect that for any real borrowed amount other than zero, the truncation shouldn't make any practical difference, but I can doublecheck.

**Review by RafilxTenfen — APPROVED** [review]

**Comment by RafilxTenfen** [comment]
LGTM ;) 

The following checks we are going to fix in another PR? 
- Protobuf / lint (pull_request) 
- Protobuf / breakage (pull_request) 
- Tests / test-unit-cover (pull_request)

**Comment by toteki** [comment]
The two protobuf ones are inevitable when we change a proto file - we don't have that pipeline working.

As for the unit test, that one's new (it just appeared as a new test was merged in from main) so I'll have to investigate and fix it.

File: x/leverage/keeper/borrows.go (hunk 12 above)
Comment (toteki) [reply x/leverage/keeper/borrows.go:25]: After checking into it:

- SetBorrow issues around zero won't be an issue, since it calls `setAdjustedBorrow(scalar * 0)` and thus doesn't leave a remainder

- For nonzero scenarios, all functions use `GetBorrow` and `SetBorrow`, all treat the truncated value as the "actual" amount, so they are not inconsistent with each other. Rounding errors never exceed one indivisible unit of token (e.g. `uatom` for Atom).

- The sum of all `AdjustedBorrow` is `TotalAdjustedBorrowed` for a denom - the `sdk.Dec` forms are always in sync (no error) but truncation could cause the sum of all `GetBorrow` to differ from `GetTotalBorrowed`. Each borrow position can contribute a maximum of 1 indivisible unit (e.g. `uatom`) to the sum of all `GetBorrow`. (Between those two approaches, `GetTotalBorrowed` is considered the correct one and has the minimum error).

- Derived quantity `BorrowUtilization` already defends against edge cases (e.g. 1000.1 / 1000) by returning a maximum of 1.0, and further quantities computed from utilization, like `LendAPY`, inherit that defense.

So at the moment, I don't see any truncation issues looming.

**Comment by toteki** [comment]
Fixed - basically they test we added from audit was minting tokens straight into the module to expand the lending pool (and thus avoid an insufficient lending pool error).

But now that uToken exchange rate is derived instantly, that mint action no longer had the desired effect, as the exchange rate increases so the total supply of uTokens is "worth" the newly expanded total module supply. This is as intended, so I just created an additional lender to lend some extra tokens and expand the lending pool without direct minting.

File: x/leverage/keeper/borrows.go (hunk 12 above)
Comment (alexanderbez) [reply x/leverage/keeper/borrows.go:25]: perfect! Noted on preference of `GetTotalBorrowed` over summing over `GetBorrow ` 👍

**Comment by alexanderbez** [comment]
Amazing work 👏
